### ISSUE TITLE
MelSpectrogram inconsistency with librosa melspectrogram

### ISSUE THREAD

**Issue by SolomidHero** [pytorch/audio#1058]
Hello! I am excited with this framework a lot and its ability to make transformations on gpu.

**Problem:**
`transforms.Spectrogram` (with power 1.) (which is real) output equals to absolute value of `librosa.stft` (which is complex) with equal parameters.

Here is spectrograms for my example audio (really close results):
<img width="686" alt="Screenshot 2020-11-26 at 22 23 12" src="https://user-images.githubusercontent.com/28161779/100386302-0aa6c180-3036-11eb-9eef-36afd6b03ed6.png">
<img width="678" alt="Screenshot 2020-11-26 at 22 24 45" src="https://user-images.githubusercontent.com/28161779/100386388-42156e00-3036-11eb-9b15-fbbbe8f79e30.png">

Next step is to get melspectrogram using `transforms.MelScale` (on `Spectrogram` with power 1) and `librosa.feature.melspectrogram` (actually power is 1., this argument not in use) (using previous spectrogram). And here we can't get the same result:
- in both steps only matmul takes place
- in `transforms.MelScale` tensors with real values multiplicated, in `librosa.feature.melspectrogram` gives us multiplication of complex based matrices, thus in the result we can get absolutely different values
- also quite misleading use of `power` in `transforms.Spectrogram` (don't need in `librosa.stft`)

And the result (differs not only in some fields, but in scale too):
<img width="679" alt="Screenshot 2020-11-26 at 22 45 46" src="https://user-images.githubusercontent.com/28161779/100387450-2fe8ff00-3039-11eb-8c78-98d2f39f6d76.png">
<img width="678" alt="Screenshot 2020-11-26 at 22 47 18" src="https://user-images.githubusercontent.com/28161779/100387552-6888d880-3039-11eb-914f-f20b06bb289c.png">

**Comment by eldrin** [comment]
Hi! I also happened to have a similar issue with MFCC. After some digging, I found librosa use `slaney` normalization for the mel-filterbank creation as the default, while torchaudio is no normalization by default. (no option to feed normalization setup at the `MelSpec`/`MelSpectrogram` construction.

https://github.com/pytorch/audio/blob/fb3ef9ba427acd7db3084f988ab55169fab14854/torchaudio/transforms.py#L243-L245

After specifying the normalization to `slaney` I got a reasonably comparable result.

![image](https://user-images.githubusercontent.com/5870216/100397839-359c1000-304c-11eb-991f-4b2cdf58bd4d.png)

It is still numerically not exactly the same, but much closer.

Right now my solution is to write some custom `MelSpec` and `MelSpectrogram` inherited from their original and adding an explicit `mel_norm` argument to pass it down to the filterbank creation :)

**Comment by SolomidHero** [comment]
But another problem is confusing use of power argument.

**Comment by mthrok** [comment]
> But another problem is confusing use of power argument.

@SolomidHero 
Can you elaborate this one? I think I have a similar opinion on the `power` used in complex values in torchaudio. I am thinking if a way to improve it while I work on the migration of native complex type.

**Comment by SolomidHero** [comment]
Sure!

Let's see a common transformation: **input (wav)** -> **melspectrogram** -> **output (same wav)** done with librosa and with torch:
```python
# some parameters:
sample_rate = 24000

stft_params = {
  'n_fft': 2048,
  'hop_length': 256,
}

# some data
y_np = np.array(...) # wav
y = torch.from_numpy(y_np)
```
Let' assume our `power` is x, that we want use in between (x is `1.` or `2.` commonly)
```python
# librosa
mel = librosa.feature.melspectrogram(y=y_np, sr=sample_rate, power=x, **stft_params) # mel_scale(S**x)
spec = librosa.feature.inverse.mel_to_stft(mel, sr=sample_rate, power=x, n_fft=stft_params['n_fft']) # S
librosa_y = librosa.griffinlim(spec, hop_length=stft_params['hop_length'])

# torch (works slower, different results with librosa, different semantic)
mel = tf.MelSpectrogram(sample_rate=sample_rate, **stft_params, power=x)(y) # mel_scale(S**x)
spec = tf.InverseMelScale(n_stft=stft_params['n_fft'] // 2 + 1, sample_rate=sample_rate)(mel)  # S**x
torch_y = tf.GriffinLim(**stft_params, power=x)(spec)
```

It is clear power is applied only to absolute value of spectrogram, and it is understandable, in my opinion, to have such parameter in `librosa.inverse.mel_to_stft`. Actually, this example already shows some inconsistency between two frameworks in semantics of arguments. Audio sound have no differ almostly.  

P.s. I made a mistake in the first comment and updated it after this one

**Comment by SolomidHero** [comment]
So, will we have an option for filter bank normalization or it is not quite necessary thing to have? Also the same question for a whole db to amplitude transformation

**Comment by mthrok** [comment]
Hi @SolomidHero @eldrin 

I looked into the difference of MelSpectrogram. As [the documentation](https://pytorch.org/audio/stable/transforms.html#torchaudio.transforms.MelSpectrogram) says, the origin of the code is from some notebooks, and it seems that there is no test to check the output values of this transform. This makes it difficult to tell if the current implementation is correct or define what is the correct behavior here.

I looked at `librosa`, hoping that they have some test to validate the resulting signal but [they do not have such test either](https://github.com/search?q=melspectrogram+repo%3Alibrosa%2Flibrosa+path%3A%2Ftests+language%3APython+language%3APython&type=Code&ref=advsearch&l=Python&l=Python).

We could change the implementations to match `librosa` but that will cause backward compatibility changes, so we really need to have a good reason to do so.  (Recently PyTorch core is doing similar move with NumPy API compatibility)

Adding an option to change normalization method for `MelSpectorgram` **completely** solve the issue discussed here?

**Comment by eldrin** [comment]
Hi @mthrok @SolomidHero 

sorry for the late response. 

> Adding an option to change normalization method for MelSpectorgram completely solve the issue discussed here?

I am afraid not. I was digging in a bit for the issue and I found it's both from the `slaney` normalization and `mel filter bank`.

`F.create_fb_matrix` function generates a mel filterbank kernel based on the formula same with the [wikipedia entry](https://en.wikipedia.org/wiki/Mel_scale). While librosa seems to offer [2 implementations](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/core/convert.py#L1143) (`htk` and `auditory toolbox`). And there is another option: normalization scheme from `slaney` or `None`. 

From the equation and the outputs, torchaudio implements [`htk` + `None`] and [`htk` + `slaney`] while librosa provides full combination of {`htk`, `auditory toolbox`} x {`None`, `slaney`}. The default setup for each packages so far is `torchaudio`: [`htk` + `None`] and `librosa`:[`auditory toolbox` + `slaney`].  

![image](https://user-images.githubusercontent.com/5870216/106395621-87259a00-6403-11eb-9fbf-40c4bef4f3de.png)

I plugged in the default librosa filterbank to the custom `MelSpectrogram` transform, and the result comply the librosa output. I measured it computing the mean squared error of resulting mel-spectrograms computed from torch audio object (using different filter banks) against the one computed from `librosa` as the reference.

```
Mean Squared Error[fb:None]: 275233.8125
Mean Squared Error[fb:slaney]: 3064.0825
Mean Squared Error[fb:librosa_auditory_toolbox+slaney]: 0.0000
```
![image](https://user-images.githubusercontent.com/5870216/106395180-0cf41600-6401-11eb-8bd8-e3626184e019.png)

Result says that once the exactly same kernel and normalization are used, the result comply the librosa output!

So eventually implementation of `auditory toolbox` based mel-filterbank is needed for the complete compliance.

[Here](https://gist.github.com/eldrin/0f8f8e5594f5dfeff5b9261e43242436)'s the code snippet for this test

**Comment by mthrok** [comment]
Hi @eldrin 

Thanks for the detailed report. This is very comprehensive.
Not having test against reference implementation is one of my concern so this will help improve the situation.
Now we understand that where the inconsistency is coming, do you think it worth adding `auditory_toolbox` implementation?

**Comment by eldrin** [comment]
Hi @mthrok 

No problem! And Yes. I think it worth several points: 1) (Although I am not 100% sure for this matter) a specific mel scale setup would benefit a particular application more. 2) many existing and new research papers and their implementations use the Slaney mel-scale setup (by choosing librosa) often, not explicitly mentioning it in their technical report or paper sometimes. 3) transitioning to torchaudio would be smoother for existing/legacy systems based on librosa

As for the [terminology](https://librosa.org/doc/main/generated/librosa.mel_frequencies.html), I find Slaney authored `auditory_toolbox,` so I think it's just `htk` implementation and `slaney` implementation for the mel-scale and normalization. I see that librosa allows any combination between two implementations at the component level.

I hope you find it useful.

**Comment by mthrok** [comment]
@eldrin @SolomidHero 

I have merged #1212 so we can pass `slaney` normalization as a parameter to `MelSpectrogram` transform. I will keep looking at a way to add other filter bank option and numerical parity to `librosa`.

**Comment by mthrok** [comment]
To summarize:

1. `torchaudio.functional.create_fb_matrix(..., norm='slaney')` is numerically compatible with `librosa.filters.mel(..., htk=True, norm="slaney")`
    * Subtask is to make `htk` option available to `create_fb_matrix` and Transforms that use this funciton.
2. `torchaudio.transforms.Spectrogram` is numerically compatible with `librosa.core.spectrum._spectrogram`. [gist](https://gist.github.com/mthrok/01f89d9bc27a7fe618bf5e8ef71b44ba)
3. Somewhere in `torchaudio.transforms.MelScale` is not matching with librosa.

**Comment by mthrok** [comment]
Okay, I did further research and could reproduce librosa's melspectrogram with torchaudio. The parameters added in #1212 helped.

### Numerical compatibility

* Spectrogram
Confirmed that [torchaudio.transforms.Spectrogram](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/transforms.py#L34-L108) and [librosa.core.spectrum._spectrogram](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/core/spectrum.py#L2434-L2525) can produce numerically comparable results. [[script](https://gist.github.com/mthrok/01f89d9bc27a7fe618bf5e8ef71b44ba#file-compare_spectrogram-py)]


```python
torchaudio_spec = torchaudio.transforms.Spectrogram(
    n_fft=n_fft,
    win_length=win_len,
    hop_length=hop_len,
    center=True,
    pad_mode="reflect",
    power=2.0,
)(waveform)
librosa_spec, _ = librosa.core.spectrum._spectrogram(
    waveform.numpy(),
    n_fft=n_fft,
    hop_length=hop_len,
    win_length=win_len,
    center=True,
    pad_mode="reflect",
    power=2.0,
)
```

![spec](https://user-images.githubusercontent.com/855818/107825892-ea8fc000-6d51-11eb-9be5-a4e31e0701ac.png)

**MSE: 5.792542556726232e-10**

* MelScale conversion
Confirmed that [torchaudio.functional.create_fb_matrix](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/functional/functional.py#L295-L357) and [librosa.filters.mel](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/filters.py#L112-L245) can produce numerically comparable results. [[script](https://gist.github.com/mthrok/01f89d9bc27a7fe618bf5e8ef71b44ba#file-compare_mel_filterbank-py)]

```python
torchaudio_mel = torchaudio.functional.create_fb_matrix(
    int(n_fft // 2 + 1),
    n_mels=n_mels,
    f_min=0.,
    f_max=sample_rate/2.,
    sample_rate=sample_rate,
    norm='slaney'
)

librosa_mel = librosa.filters.mel(
    sample_rate,
    n_fft,
    n_mels=n_mels,
    fmin=0.,
    fmax=sample_rate/2.,
    norm='slaney',
    htk=True,
).T
```

![mel_bins](https://user-images.githubusercontent.com/855818/107825904-efed0a80-6d51-11eb-8dbd-661ac01c9838.png)

**MSE: 3.6859009276685303e-16**

* MelSpectrogram
Confirmed that [torchaudio.transforms.MelSpectrogram](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/transforms.py#L400-L482) and [librosa.feature.melspectrogram](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/feature/spectral.py#L1873-L2011) can produce numerically comparable results. [[script](https://gist.github.com/mthrok/01f89d9bc27a7fe618bf5e8ef71b44ba#file-compare_melspectrogram-py)]

```python
torchaudio_melspec = torchaudio.transforms.MelSpectrogram(
    sample_rate=sample_rate,
    n_fft=n_fft,
    win_length=win_len,
    hop_length=hop_len,
    center=True,
    pad_mode="reflect",
    power=2.0,
    norm='slaney',
    onesided=True,
    n_mels=n_mels,
)(waveform)
librosa_melspec = librosa.feature.melspectrogram(
    waveform.numpy(),
    sr=sample_rate,
    n_fft=n_fft,
    hop_length=hop_len,
    win_length=win_len,
    center=True,
    pad_mode="reflect",
    power=2.0,
    n_mels=n_mels,
    norm='slaney',
    htk=True,
)
```

![mel_spec](https://user-images.githubusercontent.com/855818/107825914-f4192800-6d51-11eb-8227-28c2fb2fda46.png)

**MSE: 3.748331423025775e-09**


### Call-stacks

<details>

* MelSpectrogram
MelSpectrogram call stack of torchaudio and librosa;
    * [torchaudio.transforms.MelSpectrogram](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/transforms.py#L400-L482)
        * [torchaudio.transforms.Spectrogram](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/transforms.py#L34-L108)
            * [torchaudio.functional.spectrogram](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/functional/functional.py#L35-L107)
        * [torchaudio.transforms.MelScale](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/transforms.py#L237-L302)
            * [torchaudio.functional.create_fb_matrix](https://github.com/pytorch/audio/blob/4f9b55201dea6d0dd5165a4f684d1160ee9ca94a/torchaudio/functional/functional.py#L295-L357)
    * [librosa.feature.melspectrogram](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/feature/spectral.py#L1873-L2011)
        * [librosa.core.spectrum._spectrogram](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/core/spectrum.py#L2434-L2525)
        * [librosa.filters.mel](https://github.com/librosa/librosa/blob/a53fa56bdb6695a994008d5b6ccd0100870a6036/librosa/filters.py#L112-L245)

</details>

**Comment by expectopatronum** [comment]
Hi @mthrok, 
I also recently had the issue that my `librosa` and `torchaudio` spectrograms were not compatible. Will it be possible only if the `librosa` spectrogram is created in the same way that you showed, e.g. setting `htk=True', ...?

I am currently creating my spectrograms with very few settings:
```
librosa.feature.melspectrogram(y, n_fft=1024, hop_length=512, sr=sr)
```
and would (as a new user) expect 
```
torchaudio.transforms.MelSpectrogram(n_fft=1024, hop_length=512, sample_rate=sr)
```
in the same way. Will this be possible?
(This is mainly a problem when I try to reproduce something from a paper that used `librosa`, so I can't change how they computed the spectrograms).

Best regards
Verena

**Comment by mthrok** [comment]
Hi @expectopatronum 

Sorry for the late reply. We are revisiting #593, which adds HTK compatibility to base frequency bin generation. Once this is done, we can expose the parameter to `MelSpectrogram` and we should get the support for `htk`.

**Comment by eldrin** [comment]
Hi @mthrok 

I have double checked and from a few sources including [librosa documentation](https://librosa.org/doc/main/generated/librosa.mel_frequencies.html) and [an old email thread from auditory list](http://www.auditory.org/mhonarc/2008/msg00189.html), it looks like the one that is implemented in torchaudio is actually `htk` and the other one that is not implemented is `Slaney`.

Also I found the equation in the email thread:

```
...
p.s. here is a version of Slaney's equation, which maps the frequency
range 133 Hz to 6400 Hz to the range 0.0 to 40.0, with 1000 Hz mapping
to 13.0, and being linear below and logarithmic above that point.

mel(f) = {  (f - f_0)/f_step   for f <= f_b
         { m_b + ln(f/f_b)/m_step   for f > f_b

where f_0 = 133.33, f_step = 66.67, f_b = 1000,
m_b = (f_b - f_0)/f_sp = 13.0 (by construction)
and m_step = ln(6.4)/27 - so the range from
1000 Hz to 6400 Hz accounts for the remaining
27.0 to take the scale up to 40.
```

which seems to be the basis of the implementation of [Auditory toolbox](https://engineering.purdue.edu/~malcolm/interval/1998-010/) written by Slaney.

I hope you find it informative.

Thanks

**Comment by mthrok** [comment]
We have merged #593, which adds the equivalent of `htk=False` option of librosa. 

@eldrin The change includes roughly what you quoted, you should be able to do the comparison. Try the nightly build tomorrow and let us know if that meets your requirement.
